File: 946-ValidateStackSequences/946-ValidateStackSequences.py
from collections import deque
class Solution:
    def validateStackSequences(self, pushed: List[int], popped: List[int]) -> bool:
        popped_deck = deque(popped)
        res = []
        
        for i in pushed:
            res.append(i)

            while res and res[-1] == popped_deck[0]:
                res.pop()
                popped_deck.popleft()
        
        while res and res[-1] == popped_deck[0]:
            res.pop()
            popped_deck.popleft()

        return True if len(res) == 0 else False
        

        

File: 946-ValidateStackSequences/946-ValidateStackSequences.js
/**
 * @param {number[]} pushed
 * @param {number[]} popped
 * @return {boolean}
 */
var validateStackSequences = function(pushed, popped) {
    let res = [];
    let leftPointer = 0;

    pushed.forEach( num => {
        res.push(num);

        while(res.length > 0 && res[res.length - 1] == popped[leftPointer]){
            res.pop();
            leftPointer++;
        }
    })

    while(res.length > 0 && res[res.length - 1] == popped[leftPointer]){
        res.pop();
        leftPointer++;
    }
    
    return res.length == 0 ? true : false;
};
